### ISSUE TITLE
Add and Publish TypeScript definition with hapi package

### ISSUE THREAD

**Issue by MeirionHughes** [hapijs/hapi#3419]
I see there are two older issues related to maintaining typescript definitions; but these relate to getting them on [definitely-typed](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/hapi/index.d.ts). Recently typescript has moved forward and now supports npm packages declaring their own types. This is done via a `"typings":  "path/to/types.d.ts"` in the package.json. 

Please consider adding, maintaining and publishing typescript definitions to this (and other repos); By having it in the repo, its far easier to maintain and keep up-to-date. Potentially, this could simply be started by copying the definitely-typed definition file into here.

 Finally, uptake in [typescript is rapidly growing](http://whatpixel.com/is-typescript-worth-learning/) and reducing any conflicts between typing and run-time is probably worth the effort.

**Comment by Marsup** [comment]
> its far easier to maintain and keep up-to-date

As far as I know, no one in this organization is using typescript daily or at all. Those files will likely be either stale or wrong very quickly.

**Comment by vdeturckheim** [comment]
Before anything, I am a happy TypeScript user in some fields and I know how useful are typings files.
However I do not feel the hapi team has to create such file: it would be a lot of work and a commitment to maintain this as the framework evolves. hapi is a Node.js framework written in JavaScript and is aimed at being used with this language.

If someone commits to maintain it for the next few years, I might agree. But atm. I do not think it should happen.

**Comment by nlf** [comment]
Agreed. The typings that exist are already pretty outdated, so would need to be reviewed for correctness and updated in the first place on top of maintained for changes across multiple projects. This is a *very* significant undertaking. Unless someone who already knows and uses typescript and is deeply familiar with hapi volunteers, I don't see this happening any time soon.

**Comment by MeirionHughes** [comment]
Okaydoky; having looked at the typings I agree it would require a lot of effort to maintain it manually.

**Comment by AJamesPhillips** [comment]
Thanks for the comments maintainers and thanks for the excellent HapiJS.

@MeirionHughes I don't know if you've seen this: https://www.npmjs.com/package/@types/hapi  It should allow for using the .d.ts and for saving the hapi team from the noise that goes along with keeping it up to date.

**Comment by corbinu** [comment]
@AJamesPhillips The @types modules are built from DefinitelyTyped. 

I have been slowly updating the hapi types an a adding new for hapi subprojects like lab as we roll all our code to TS once the roll is complete and I am sure they are correct I will PR there and maintain. That should be the best solution for everybody I think.

**Comment by AJamesPhillips** [comment]
@corbinu I was going to **attempt** to update the hapi js .d.ts to version 16.1 .  I tried to find your work on your repo at git@github.com:corbinu/DefinitelyTyped.git but my git fu hasn't succeeded so far.  Could you point me towards a commit/branch with some of your recent work on it.  Thank you very much.

**Comment by corbinu** [comment]
@AJamesPhillips The types aren't finished yet. Honestly not sure if I am going to bother anymore as it is such a small part of my app that is not typed and the API is the easiest to smoke test. So just going to use it as an "any"

**Comment by AJamesPhillips** [comment]
I've completed updating the types for hapi 16.1.1 and some of the plugins.  They're now available here:  https://github.com/AJamesPhillips/DefinitelyTyped/tree/hapi-v16.1/types/hapi (commit: https://github.com/AJamesPhillips/DefinitelyTyped/commit/97abba2d62da3f6f659330f5a0c118c78e9b98a4 )
I made a mistake whilst updating the inert plugin types before: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/15616/commits/b33942b17b4117f2fa5b2eae6ad3baa55323768f so I'm only going to submit these new types to the repository when I have tested them for a week or so.  If anyone else who uses Hapi with TypeScript could test these with real production code and submit any improvements that'd obviously be very useful.
